File: insertionSort.js
/*
"user strict";

//When N is guaranteed to be small,
//including as the base case of a quick sort or merge sort.
//While this is O(N^2), it has a very small constant and is a stable sort.

var dataArray = [8, 5, 3, 12];


function prevElem_GTE_toInsert(currentValue, toInsertValue) {
  return (currentValue >= toInsertValue);
}

function findInsertionPoint(array, outerIndex) {
    var valueToInsert = array[outerIndex];
    var innerIndex = outerIndex;

    console.log("Find niche to insert: " + valueToInsert);


    while(innerIndex > 0 && prevElem_GTE_toInsert(array[innerIndex-1], valueToInsert)) {
        array[innerIndex] = array[innerIndex-1];
        innerIndex--;
    }

    console.log("niche found at: " + innerIndex + ", for value: " + valueToInsert);
    array[innerIndex] = valueToInsert;
}


function insertionSort(unsortedArray) {

      var data = unsortedArray.slice(0);

      for (var outerIndex = 1; outerIndex <= data.length-1; outerIndex++) {
        console.log("==== outerIndex: "+outerIndex+" ====");
        findInsertionPoint(data, outerIndex);
      }
      return data;
}

let sorted = insertionSort(dataArray);
console.log(sorted);
*/


// testing 1 2 3
function subsetSum(numbers, target, partial) {
  var s, n, remaining;

  partial = partial || [];

  // sum partial
  s = partial.reduce(function (a, b) {
    return a + b;
  }, 0);

  // check if the partial sum is equals to target
  if (s === target) {
    console.log("%s=%s", partial.join("+"), target)
  }

  if (s >= target) {
    return;  // if we reach the number why bother to continue
  }

  for (var i = 0; i < numbers.length; i++) {
    n = numbers[i];
    remaining = numbers.slice(i + 1);
    subsetSum(remaining, target, partial.concat([n]));
  }
}

subsetSum([3, 3, 3, 3, 4, 4, 4, 6, 6, 8, 12],12);

File: insertionSort.cpp
#include <iostream>

using namespace std;

// feature 1
// feature 2
// feature 3
bool prevElem_GTE_toInsert(int currentValue, int toInsertValue) {
    return (currentValue >= toInsertValue);
}

void findInsertionPoint(int arr [], int outerIndex) {
    
    
    int valueToInsert = arr[outerIndex];
    int innerIndex = outerIndex;
    cout << "\nFind niche to insert value " << valueToInsert << endl;
    
    while (innerIndex > 0 &&
           prevElem_GTE_toInsert(arr[innerIndex-1], valueToInsert)) {
        arr[innerIndex] = arr[innerIndex-1];
        innerIndex--;
    }
    
    cout << "\nniche found at: " << innerIndex << ", for value " << valueToInsert;
    arr[innerIndex] = valueToInsert;
}


void insertionSort(int arr [], int size) {
    
    cout << "\nSorting an array that is: " << size << endl;
    for ( int outerIndex = 1; outerIndex <= size-1; outerIndex++) {
        cout << "\n=== outerIndex: " << outerIndex << " ====" << endl;
        findInsertionPoint(arr, outerIndex);
    }
}



int main() {

    int arr [] = {387, 3,1, 7,100, 2,88, 9,1};
    int size = sizeof(arr) / sizeof(int);
    
    insertionSort(arr, size);
    
    cout << "\n---- result ----" << endl;
    for ( int i = 0; i < size; i++) {
        cout << arr[i] << endl;
    }
    return 0;
}
